Mathlib/RingTheory/Norm/Transitivity.lean

```diff
@@ -0,0 +1,195 @@
+Authors: Junyan Xu
+-/
+import Mathlib.LinearAlgebra.Matrix.Block
+import Mathlib.LinearAlgebra.Matrix.Charpoly.Coeff
+import Mathlib.RingTheory.Norm.Defs
+import Mathlib.RingTheory.PolynomialAlgebra
+
+/-!
+# Transitivity of algebra norm
+
+Suppose we have an `R`-algebra `S` with a finite basis. For each `s : S`,
+the determinant of the linear map given by multiplying by `s` gives information
+about the roots of the minimal polynomial of `s` over `R`.
+
+## References
+
+ * [silvester2000] Silvester, *Determinants of Block Matrices*, The Mathematical Gazette (2000).
+
+-/
+
+variable {R S A n m : Type*} [CommRing R] [CommRing S]
+variable (M : Matrix m m S) [DecidableEq m] [DecidableEq n] (k : m)
+open Matrix Polynomial
+
+namespace Algebra.Norm.Transitivity
+
+/-- Given a ((m-1)+1)x((m-1)+1) block matrix [[A,b],[c,d]] `M`, `auxMat M k` is the auxiliary matrix
+[[dI,0],[-c,1]]. `k` corresponds to the last row/column of the matrix. -/
+def auxMat : Matrix m m S :=
+  of fun i j ↦
+    if j = k then
+      if i = k then 1 else 0
+    else if i = k then -M k j
+    else if i = j then M k k
+    else 0
+
+/-- `aux M k` is lower triangular. -/
+lemma auxMat_blockTriangular : (auxMat M k).BlockTriangular (· ≠ k) :=
+  fun i j lt ↦ by
+    simp_rw [lt_iff_not_le, le_Prop_eq, Classical.not_imp, not_not] at lt
+    rw [auxMat, of_apply, if_pos lt.2, if_neg lt.1]
+
+lemma auxMat_toSquareBlock_true : (auxMat M k).toSquareBlock (· ≠ k) True = M k k • 1 := by
+  ext i j
+  simp [auxMat, toSquareBlock_def, if_neg (of_eq_true i.2), if_neg (of_eq_true j.2),
+    Matrix.one_apply, Subtype.ext_iff]
+
+lemma auxMat_toSquareBlock_false : (auxMat M k).toSquareBlock (· ≠ k) False = 1 := by
+  ext ⟨i, hi⟩ ⟨j, hj⟩
+  rw [eq_iff_iff, iff_false, not_not] at hi hj
+  simp [auxMat, toSquareBlock_def, if_pos hi, if_pos hj, Matrix.one_apply, if_pos (hj ▸ hi)]
+
+variable [Fintype m]
+
+/-- `M * aux M k` is upper triangular. -/
+lemma mul_auxMat_blockTriangular : (M * auxMat M k).BlockTriangular (· = k) :=
+  fun i j lt ↦ by
+    simp_rw [lt_iff_not_le, le_Prop_eq, Classical.not_imp] at lt
+    simp_rw [Matrix.mul_apply, auxMat,  of_apply, if_neg lt.2, mul_ite, mul_neg, mul_zero]
+    rw [Finset.sum_ite, Finset.filter_eq', if_pos (Finset.mem_univ _), Finset.sum_singleton,
+      Finset.sum_ite_eq', if_pos, lt.1, mul_comm, neg_add_cancel]
+    exact Finset.mem_filter.mpr ⟨Finset.mem_univ _, lt.2⟩
+
+/-- The lower-right corner of `M * aux M k` is the same as the corner of `M`. -/
+lemma mul_auxMat_corner : (M * auxMat M k) k k = M k k := by simp [Matrix.mul_apply, auxMat]
+
+lemma mul_auxMat_toSquareBlock_true :
+    (M * auxMat M k).toSquareBlock (· = k) True = M k k • 1 := by
+  ext ⟨i, hi⟩ ⟨j, hj⟩
+  rw [eq_iff_iff, iff_true] at hi hj
+  simp [toSquareBlock_def, hi, hj, mul_auxMat_corner]
+
+/-- The upper-left block of `M * aux M k`. -/
+def mulAuxMatBlock : Matrix { a // (a = k) = False } { a // (a = k) = False } S :=
```

Comment:
Possibly worth a remark about the weird type here? Do you get the impression that the API for block matrices is wrong?

Reply:
I've now made it a notation so we don't need to specify the type. I think the API is easy to use, and these `= True` `= False` are unavoidable if we use a predicate to specify the blocks. Even though `rw [eq_iff_iff, iff_false, not_not]` is slightly annoying, I think reindexing with `{ a // a = k }` her would make things more complicated.